### PR TITLE
Fix span error tagging in GO YARPC/tchannel

### PR DESCRIPTION
**Pull request by kexiongliu123**
- Fix span error tagging in GO YARPC/tchannel
- inbound and outbound traffic
- unit tests

### CODE REVIEW

File: transport/tchannel/channel_outbound.go [hunk 1]
```diff
@@ -37,6 +39,11 @@ var (
 	_ introspection.IntrospectableOutbound = (*ChannelOutbound)(nil)
 )
 
+const (
+	//TracingTagStatusCode is the span tag key for the YAPRC status code.
+	TracingTagStatusCode = "rpc.yarpc.status_code"
```
Comment (biosvs) [inline transport/tchannel/channel_outbound.go:44]: How many copies of this tag do we have at this point? Let's extract it to some package and reuse.

File: transport/tchannel/channel_outbound.go [hunk 2]
```diff
@@ -22,6 +22,8 @@ package tchannel
 
 import (
 	"context"
+	"github.com/opentracing/opentracing-go"
```
Comment (biosvs) [inline transport/tchannel/channel_outbound.go:25]: Please sort imports.

File: transport/tchannel/channel_outbound.go [hunk 3]
```diff
@@ -103,10 +110,15 @@ func (o *ChannelOutbound) IsRunning() bool {
 
 // Call sends an RPC over this TChannel outbound.
 func (o *ChannelOutbound) Call(ctx context.Context, req *transport.Request) (*transport.Response, error) {
+	span := opentracing.SpanFromContext(ctx)
+	if span != nil {
+		defer span.Finish()
```
Comment (biosvs) [inline transport/tchannel/channel_outbound.go:115]: Why we're finishing span in a place where we didn't start it?

File: transport/tchannel/channel_outbound.go [hunk 4]
```diff
@@ -103,10 +110,15 @@ func (o *ChannelOutbound) IsRunning() bool {
...
+	}
 	if req == nil {
 		return nil, yarpcerrors.InvalidArgumentErrorf("request for tchannel channel outbound was nil")
 	}
 	if err := o.once.WaitUntilRunning(ctx); err != nil {
+		UpdateSpanWithErr(span, err, yarpcerrors.FromError(err).Code())
```
Comment (biosvs) [inline transport/tchannel/channel_outbound.go:121]: Let's reuse yarpcerrors.FromError(err) result.

File: transport/tchannel/channel_outbound.go [hunk 5]
```diff
@@ -151,6 +163,7 @@ func (o *ChannelOutbound) Call(ctx context.Context, req *transport.Request) (*tr
 	}
 
 	if err != nil {
+		UpdateSpanWithErr(span, err, yarpcerrors.FromError(err).Code())
```
Comment (biosvs) [inline transport/tchannel/channel_outbound.go:166]: Shouldn't toYARPCError(req, err) result be used here?

File: transport/tchannel/handler.go [hunk 6]
```diff
@@ -183,6 +190,7 @@ func (h handler) callHandler(ctx context.Context, call inboundCall, responseWrit
 
 	ctx, headers, err := readRequestHeaders(ctx, call.Format(), call.Arg2Reader)
 	if err != nil {
+		UpdateSpanWithErr(span, err, yarpcerrors.FromError(err).Code())
```
Comment (biosvs) [inline transport/tchannel/handler.go:193]: Same, yarpcerrors created later in RequestHeadersDecodeError

File: transport/tchannel/handler.go [hunk 7]
```diff
@@ -201,25 +209,30 @@ func (h handler) callHandler(ctx context.Context, call inboundCall, responseWrit
 
 	body, err := call.Arg3Reader()
 	if err != nil {
+		UpdateSpanWithErr(span, err, yarpcerrors.FromError(err).Code())
```
Comment (biosvs) [inline transport/tchannel/handler.go:212]: I though UpdateSpanWithErr returns err and so it can be used as `return UpdateSpanWithErr(...`?

**Review by biosvs — CHANGES_REQUESTED** [review]

File: transport/tchannel/handler.go [hunk 8]
```diff
@@ -116,12 +116,16 @@ func (h handler) handle(ctx context.Context, call inboundCall) {
 	responseWriter := h.newResponseWriter(call.Response(), call.Format(), h.headerCase)
 	defer responseWriter.ReleaseBuffer()
 
+	span := opentracing.SpanFromContext(ctx)
```
Comment (ChenX1993) [inline transport/tchannel/handler.go:119]: The span here is not the right tchannel inbound span.
The tchannel inbound span is available after `tchannel.ExtractInboundSpan(...)` is called in `callHandler` method, and you can get the span from the context.

File: transport/tchannel/channel_outbound.go [hunk 9]
```diff
@@ -103,10 +110,15 @@ func (o *ChannelOutbound) IsRunning() bool {
 
 // Call sends an RPC over this TChannel outbound.
 func (o *ChannelOutbound) Call(ctx context.Context, req *transport.Request) (*transport.Response, error) {
+	span := opentracing.SpanFromContext(ctx)
```
Comment (ChenX1993) [inline transport/tchannel/channel_outbound.go:113]: This is not the right tchannel outbound span.
The tchannel outbound span is created when`xxx.BeginCall(..)` is invoked, and the [span is created in tchannel-go](https://github.com/uber/tchannel-go/blob/dev/outbound.go#L120).

It seems that tchannel doesn't expose the outbound span currently. We probably need to expose it from tchannel-go first.

Comment (kexiongliu123) [reply transport/tchannel/channel_outbound.go:113]: I agree. I will make this PR solely deal with inbound error tagging. For outbound, I will implement after exposing it from tchannel-go.

File: transport/tchannel/handler.go [hunk 10]
```diff
@@ -195,31 +195,36 @@ func (h handler) callHandler(ctx context.Context, call inboundCall, responseWrit
 		tracer := h.tracer
 		ctx = tchannel.ExtractInboundSpan(ctx, tcall.InboundCall, headers.Items(), tracer)
 	}
-
+	span := opentracing.SpanFromContext(ctx)
```
Comment (biosvs) [inline transport/tchannel/handler.go:198]: Probably, I'm loosing the idea of changes.

There's a function called `handle` (not `Handle`), why it should not have code like

```
extractOpenTracingSpan := &transport.ExtractOpenTracingSpan{
		ParentSpanContext: parentSpanCtx,
		Tracer:            tracer,
		TransportName:     TransportName,
		StartTime:         start,
		ExtraTags:         yarpc.OpentracingTags,
	}
	ctx, span := extractOpenTracingSpan.Do(ctx, transportRequest)
	defer span.Finish()
```

And I'd expect in this case span should be updated only once - based on the error returned from the `callHandler` func.

**Pull request closed without merging**